exporter/datadogexporter/internal/metrics/transform.go

```diff
@@ -0,0 +1,208 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metrics // import "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter/internal/metrics"
+
+import (
+	"context"
+	"fmt"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/confmap"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"gopkg.in/yaml.v2"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/metricstransformprocessor"
+)
+
+// Transformer is able to transform a set of metrics by either modifying or adding
+// new ones.
+type Transformer struct {
+	names    map[string]struct{}
+	consumer func(ctx context.Context, md pmetric.Metrics) error
+}
+
+// Transform applies a set of changes to the given set of metrics.
+func (t *Transformer) Transform(ctx context.Context, md pmetric.Metrics) error {
+	return t.consumer(ctx, md)
+}
+
+// Has reports whether the metric with the given name was or would be added by the
+// transformer.
+func (t *Transformer) Has(name string) bool {
+	_, ok := t.names[name]
+	return ok
+}
+
+// NewInfraTransformFunc returns a new TransformFunc which converts OpenTelemetry system and process metrics
+// to Datadog compatible ones.
+func NewInfraTransformer(ctx context.Context, set component.ExporterCreateSettings) (*Transformer, error) {
+	cfg, err := yamlToConfig()
+	if err != nil {
+		return nil, fmt.Errorf("error unmarshalling yaml to config: %w", err)
+	}
+	pset := component.ProcessorCreateSettings(set)
+	noop, err := consumer.NewMetrics(
+		consumer.ConsumeMetricsFunc(
+			func(_ context.Context, _ pmetric.Metrics) error { return nil },
+		),
+	)
+	if err != nil {
+		return nil, err
+	}
+	proc, err := metricstransformprocessor.NewFactory().CreateMetricsProcessor(ctx, pset, cfg, noop)
+	if err != nil {
+		return nil, err
+	}
+	transformer := Transformer{
+		names:    make(map[string]struct{}),
+		consumer: proc.ConsumeMetrics,
+	}
+	for _, t := range cfg.Transforms {
+		if t.NewName != "" {
+			transformer.names[t.NewName] = struct{}{}
+		}
+	}
+	return &transformer, err
+}
+
+func yamlToConfig() (*metricstransformprocessor.Config, error) {
+	var raw map[string]interface{}
+	if err := yaml.Unmarshal(transformYaml, &raw); err != nil {
+		return nil, err
+	}
+	cid := config.NewComponentIDWithName(config.MetricsDataType, "dd-infra-metricstransformer")
```

Comment:
I would rather make this component id be constructed from the exporter id (you can get that from the `ExporterSettings` struct), something like:

```suggestion
	cid := config.NewComponentIDWithName(config.MetricsDataType, exporterID + "/infra-metricstransformer")
```

If someone has two Datadog exporters they would end up building two processors with the same ID. This can lead to trouble (e.g. internal telemetry metrics could become clobbered), and two components of the same kind having the same ID is something that never happens, so I am worried that other bad things could happen, either now or at some point in the future.

Reply:
Makes sense. Done. Please check latest commit. I hope that's what you were implying.